spacy/matcher/matcher.pyx

```diff
@@ -867,21 +867,34 @@ class _SetPredicate:
 
     def __call__(self, Token token):
         if self.is_extension:
-            value = get_string_id(token._.get(self.attr))
+            value = token._.get(self.attr)
         else:
             value = get_token_attr_for_matcher(token.c, self.attr)
 
-        if self.predicate in ("IS_SUBSET", "IS_SUPERSET", "INTERSECTS"):
-            if self.attr == MORPH:
-                # break up MORPH into individual Feat=Val values
-                value = set(get_string_id(v) for v in MorphAnalysis.from_id(self.vocab, value))
+        unexpected_value_type = False
+        if self.predicate in ("IN", "NOT_IN"):
+            if isinstance(value, (str, int)):
+                value = get_string_id(value)
             else:
-                # treat a single value as a list
-                if isinstance(value, (str, int)):
-                    value = set([get_string_id(value)])
+                unexpected_value_type = True
+        elif self.predicate in ("IS_SUBSET", "IS_SUPERSET", "INTERSECTS"):
+            # ensure that all values are enclosed in a set
+            if self.attr == MORPH:
+                if isinstance(value, int):
+                    # break up MORPH into individual Feat=Val values
+                    value = set((get_string_id(v) for v in MorphAnalysis.from_id(self.vocab, value)))
                 else:
-                    value = set(get_string_id(v) for v in value)
-        if self.predicate == "IN":
+                    unexpected_value_type = True
+            elif isinstance(value, Iterable):
+                value = set(get_string_id(v) for v in value)
```

Comment:
Couldn't this `Iterable` contain something that you can't convert?

In general I think that this is getting too convoluted and we can shortcut to `False` much earlier when getting the value instead, something like:

```
        if self.is_extension:
            value = token._.get(self.attr)
            if isinstance(value, (str, int)):
                value = get_string_id(value)
            else:
                if not all(isinstance(v, (str, int)) for v in value):
                    return False
                value = set(get_string_id(v) for v in value)
```

and then add tests like this if there aren't any already:

```
matcher = Matcher(en_vocab)
Token.set_extension("ext", default=[])
pattern = [{"_": {"ext": {"INTERSECTS": ["A", "C"]}}}]
matcher.add("M", [pattern])
doc = Doc(en_vocab, words=["a", "b", "c"])
doc[0]._.ext = ["A", "B"]
assert len(matcher(doc)) == 1
```

and

```
matcher = Matcher(en_vocab)
Token.set_extension("ext", default=[])
pattern = [{"_": {"ext": {"INTERSECTS": ["A", "C"]}}}]
matcher.add("M", [pattern])
doc = Doc(en_vocab, words=["a", "b", "c"])
doc[0]._.ext = [["A"], "B"]
assert len(matcher(doc)) == 0
```

I think the tests can go into the existing operator tests?

And it's still possible there's some edge case for `value` vs. `set(values)` that I've missed. The main edge case I'd be concerned about is iterating over a string value into a set of individual chars by accident, but I think that's covered?

Reply:
Hmm, the above suggestion won't work as we still need to gate the conversion of `Iterable` extension values for the IN and NOT IN predicates (otherwise the unhashable resultant set will raise an error downstream):
```
if self.is_extension:
            value = token._.get(self.attr)
            if isinstance(value, (str, int)):
                value = get_string_id(value)
            else:
                if not all(isinstance(v, (str, int)) for v in value):
                    return False
                value = set(get_string_id(v) for v in value) # This cannot be done for IN and NOT_IN
```
We could expand the `if not all...` statement to include a check for the above predicates, but that seems less readable to me than the current proposal, more so given that we check for the remaining predicates immediately afterwards to do more special handling of `MORPH` attributes. Also, if we add new attributes in the future whose values aren't `str` or `int`, we'll need to move the type-checking to the outer scope all the same.

> The main edge case I'd be concerned about is iterating over a string value into a set of individual chars by accident, but I think that's covered?

Yeah, that shouldn't happen as in the current implementation.

> Couldn't this Iterable contain something that you can't convert?

Good point. The `get_string_id` call will raise an error in that case, but we'll want to simply return `False` there too. I can expand the `if` statement to short-circuit on a failed convertibility check.

Reply:
Ah, that makes sense. In general, I still find this a bit hard to follow, and I think it would be simpler to just `return False` rather than adding the variable `unexpected_value_type`?

And would it be possible to refactor how `self.value` is stored based on the operator so that there's no difference in the `value` conversion here? (Sets of `frozenset` or something like that?)

Reply:
> Ah, that makes sense. In general, I still find this a bit hard to follow, and I think it would be simpler to just return False rather than adding the variable unexpected_value_type?

Sure.

> And would it be possible to refactor how self.value is stored based on the operator so that there's no difference in the value conversion here? (Sets of frozenset or something like that?)

I could be misunderstanding you, but do you mean flipping the conversion logic so that we do the `self.predicate` checks in the ctor instead (and use a single statement to normalize `value` in `__call__`)? 

I don't see how we can do that as we need to normalize the per-`Token` attribute values to the same form as those we're matching against. So, we'll need to keep the conversions in sync with each other. Also, the type of `value` is less constrained than that of `self.value`, which has to be an `Iterable` - this means we'll always need more logic to handle all the cases.